kbpagesconfig/editor.go

```diff
@@ -0,0 +1,206 @@
+	f, err := os.Open(kbpConfigPath)
+	switch {
+	case err == nil:
+		var cfg config.Config
+		cfg, editor.originalConfigStr, err = readConfig(f)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"reading config file %s error: %v", kbpConfigPath, err)
+		}
+		if cfg.Version() != config.Version1 {
+			return nil, fmt.Errorf(
+				"unsupported config version %s", cfg.Version())
+		}
+		editor.kbpConfig = cfg.(*config.V1)
+	case os.IsNotExist(err):
+		editor.kbpConfig = config.DefaultV1()
+	default:
+		return nil, fmt.Errorf(
+			"open file %s error: %v", kbpConfigPath, err)
+	}
+	return editor, nil
+}
+
+func (e *kbpConfigEditor) confirmAndWrite() error {
+	buf := &bytes.Buffer{}
+	if err := e.kbpConfig.Encode(buf, true); err != nil {
+		return fmt.Errorf("encoding config error: %v", err)
+	}
+	newConfigStr := buf.String()
+	if newConfigStr == e.originalConfigStr {
+		fmt.Println("no change is made to the config")
+		return nil
+	}
+
+	// print the diff
+	d := diffmatchpatch.New()
+	fmt.Println(
+		d.DiffPrettyText(
+			d.DiffMain(e.originalConfigStr, newConfigStr, true)))
+
+	// ask user to confirm
+	input, err := term.Prompt(fmt.Sprintf(
+		"confirm writing above changes to %s? (y/N): ", e.kbpConfigPath))
+	if err != nil {
+		return fmt.Errorf("getting confirmation error: %v", err)
+	}
+	if strings.ToLower(input) != "y" {
+		return fmt.Errorf("write not confirmed")
+	}
+
+	// write the new config to kbpConfigPath
+	f, err := os.Create(e.kbpConfigPath)
+	if err != nil {
+		return fmt.Errorf(
+			"opening file [%s] error: %v", e.kbpConfigPath, err)
+	}
+	if _, err = f.WriteString(newConfigStr); err != nil {
+		return fmt.Errorf(
+			"writing config to file [%s] error: %v", e.kbpConfigPath, err)
+	}
+	if err = f.Close(); err != nil {
+		return fmt.Errorf(
+			"closing file [%s] error: %v", e.kbpConfigPath, err)
+	}
+
+	return nil
+}
+
+func (e *kbpConfigEditor) addUser(username string) error {
+	if _, ok := e.kbpConfig.Users[username]; ok {
+		return fmt.Errorf("user %s already exists", username)
+	}
+	input, err := term.PromptPassword(fmt.Sprintf(
+		"enter a password for %s: ", username))
+	if err != nil {
+		return fmt.Errorf("getting password error: %v", err)
+	}
+	password := strings.TrimSpace(input)
+	if len(password) == 0 {
+		return fmt.Errorf("empty password")
+	}
+	hashed, err := bcrypt.GenerateFromPassword(
+		[]byte(password), bcrypt.DefaultCost)
+	if err != nil {
+		return fmt.Errorf("hashing password error: %v", err)
+	}
+	if e.kbpConfig.Users == nil {
+		e.kbpConfig.Users = make(map[string]string)
+	}
+	e.kbpConfig.Users[username] = string(hashed)
+	return nil
+}
+
+func (e *kbpConfigEditor) removeUser(username string) {
+	delete(e.kbpConfig.Users, username)
+}
+
+func (e *kbpConfigEditor) setAnonymousPermission(
+	permsStr string, pathStr string) error {
+	if e.kbpConfig.ACLs == nil {
+		e.kbpConfig.ACLs = make(map[string]config.AccessControlV1)
+	}
+	pathACL := e.kbpConfig.ACLs[pathStr] // struct
```

Comment:
What does the comment mean?

Reply:
It means pathACL is a struct which is pass-by-value. Seems like it's causing confusion more than helpful. Will remove.